File: Medium/328. Odd Even Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        ListNode* temp = head;
        ListNode* evenH = new ListNode(0);
        ListNode* evenC = evenH;
        while(temp){
            evenC -> next = temp -> next;
            evenC = evenC -> next;
            if(evenC){
                temp -> next = evenC -> next;
            }
            if(!temp -> next){
                temp -> next = evenH -> next;
                break;
            }
            temp = temp -> next;
        }
        return head;
    }
};

File: Java/Medium/328. Odd Even Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        
        ListNode o = head;
        if (head == null) {
            return head;
        }
        ListNode e = head.next;
        ListNode ot = o;
        ListNode et = e;
        if (e == null) {
            return head;
        }
        head.next = null;
        boolean isEven = false;
        ListNode temp = e.next;
        e.next = null;
        while(temp != null) {
            ListNode next = temp.next;
            temp.next = null;
            if (isEven) {
                et.next = temp;
                et = et.next;
            } else {
                ot.next = temp;
                ot = ot.next;
            }
            isEven = !isEven;
            temp = next;
        }
        ot.next = e;
        return head;
    }
}
